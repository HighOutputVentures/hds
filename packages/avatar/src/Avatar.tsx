import { Avatar, AvatarBadge, Box, Icon, useMultiStyleConfig } from "@chakra-ui/react";
import { SystemStyleObject } from "@chakra-ui/styled-system";
import * as React from "react";
import UserIcon from "./icons/UserIcon";
import VerifiedIcon from "./icons/VerifiedIcon";
import { Badgeable, Clickable, ResponsiveSize } from "./types";
import { omit } from "./utils";

type AvatarSize = "xs" | "sm" | "md" | "lg" | "xl" | "2xl" | "3xl" | "4xl" | "5xl";

export type AvatarBaseProps = {
  src?: string;
  name?: string;
  size?: AvatarSize | ResponsiveSize<AvatarSize>;
  online?: boolean;
  onlineIndicator?: boolean;
  verified?: boolean;
  /* ðŸ”’ internals  */
  __elevated?: boolean;
  __bordered?: boolean;
  __testId?: string;
};

export type AvatarProps = SystemStyleObject & AvatarBaseProps & Badgeable & Clickable;

/**
 *
 * ### HDS Avatar
 *
 * ðŸ’¡
 *
 * Please only use the `size`s `2xl` and below.
 * `3xl` and up are only used for profile internally
 * and will be removed in the future.
 * They also does not support `badge` and other stuffs
 *
 */
export default function HdsAvatar(props: AvatarProps) {
  const { src, name, size = "md", onlineIndicator, online, __bordered, __elevated,__testId, ...p } = props;

  const styles = useMultiStyleConfig("Avatar", { size, variant: "hds" });

  return (
    <Avatar
      variant="hds"
      src={src}
      size={size}
      name={name}
      icon={<Icon as={UserIcon} color="#475467" />}
      data-online={!!online}
      data-fallback={!src}
      data-testid={__testId ?? "hds-avatar"}
      data-elevated={!!__elevated}
      data-bordered={!!__bordered}
      data-clickable={!!p.clickable}
      {...(p.clickable && {
        tabIndex: 0,
        onClick: p.onClick,
      })}
      sx={{
        ...omit(
          p,
          "badge",
          "clickable",
          /* @ts-ignore "Include types which utilizes type guard-ing." */
          "onClick",
          "badgeIcon",
        ),
      }}
    >
      {!!onlineIndicator /* <!-- Online Indicator --> */ && (
        <AvatarBadge role="presentation" aria-label={online ? "Online" : "Offline"} />
      )}

      {!!p.verified /* <!-- Verified --> */ && (
        <Icon as={VerifiedIcon} role="presentation" aria-label="Verified" sx={styles.verified} />
      )}

      {!!p.badge /* <!-- Badge --> */ && (
        <Box role="presentation" aria-label="Company Icon" sx={styles.customIcon}>
          <Icon as={p.badgeIcon} />
        </Box>
      )}
    </Avatar>
  );
}
