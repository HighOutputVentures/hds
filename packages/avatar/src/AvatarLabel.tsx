import { Box, Text } from "@chakra-ui/react";
import * as React from "react";
import Avatar, { AvatarProps } from "./Avatar";

type Size = "sm" | "md" | "lg" | "xl";

export interface AvatarLabelProps extends AvatarProps {
  size?: Size;
  supportText?: string;
}

export default function AvatarLabel(props: AvatarLabelProps) {
  const { size = "md", name, supportText, ...others } = props;

  return (
    <Box
      sx={{
        display: "flex",
        alignItems: "center",
        justifyContent: "center",

        ...(size === "sm" && { gap: "10px" }),
        ...(size === "md" && { gap: "12px" }),
        ...(size === "lg" && { gap: "12px" }),
        ...(size === "xl" && { gap: "16px" }),
      }}
      data-testid="hds.avatar-label"
    >
      <Avatar name={name} size={size} {...others} />

      <Box>
        <Text
          sx={{
            color: "neutrals.800",

            ...(size === "sm" && { fontSize: "14px", lineHeight: "14px" }),
            ...(size === "md" && { fontSize: "14px", lineHeight: "14px" }),
            ...(size === "lg" && { fontSize: "18px", lineHeight: "18px" }),
            ...(size === "xl" && { fontSize: "20px", lineHeight: "20px" }),
          }}
          data-testid="hds.avatar-label.name"
        >
          {name}
        </Text>

        <Text
          sx={{
            color: "neutrals.600",
            marginTop: "4px",

            ...(size === "sm" && { fontSize: "12px", lineHeight: "12px" }),
            ...(size === "md" && { fontSize: "14px", lineHeight: "14px" }),
            ...(size === "lg" && { fontSize: "14px", lineHeight: "14px" }),
            ...(size === "xl" && { fontSize: "18px", lineHeight: "18px" }),
          }}
          data-testid="hds.avatar-label.support-text"
        >
          {supportText}
        </Text>
      </Box>
    </Box>
  );
}
