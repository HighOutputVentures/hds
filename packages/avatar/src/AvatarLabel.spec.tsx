import { render, screen } from "@testing-library/react";
import * as React from "react";
import renderer from "react-test-renderer";
import Avatar from "./Avatar";
import AvatarLabel from "./AvatarLabel";

describe("AvatarLabel", () => {
  beforeEach(() => {
    render(
      <AvatarLabel src="" name="John Doe" supportText="johndoe@dummy.blaa">
        <Avatar />
      </AvatarLabel>,
    );
  });

  it("Should render avatar", () => {
    expect(screen.queryByRole("img", { name: "avatar" })).toBeDefined();
  });

  it("Should render name", () => {
    expect(screen.queryByText("John Doe")).toBeDefined();
  });

  it("Should render support text", () => {
    expect(screen.queryByText("johndoe@dummy.blaa")).toBeDefined();
  });

  describe("Snapshot", () => {
    it("Should match snapshot", () => {
      const component = renderer.create(
        <AvatarLabel src="" name="John Doe" supportText="johndoe@dummy.blaa">
          <Avatar />
        </AvatarLabel>,
      );

      const tree = component.toJSON();
      expect(tree).toMatchSnapshot();
      component.unmount();
    });
  });
});
