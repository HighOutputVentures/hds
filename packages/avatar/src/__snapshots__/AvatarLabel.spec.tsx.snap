// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AvatarLabel Snapshot Should match snapshot 1`] = `
<div
  className="css-0"
>
  <div
    className="css-0"
  >
    <span
      className="chakra-avatar css-1ruhx4w"
      data-bordered={false}
      data-clickable={false}
      data-elevated={false}
      data-fallback={true}
      data-online={false}
      data-testid="hds-avatar"
    >
      <div
        aria-label="John Doe"
        className="chakra-avatar__initials css-0"
        role="img"
      >
        JD
      </div>
    </span>
  </div>
  <div
    className="css-0"
    data-testid="hds.avatar-label"
  >
    <p
      className="chakra-text css-luvoqg"
    >
      John Doe
    </p>
    <p
      className="chakra-text css-1hu2wkm"
    >
      johndoe@dummy.blaa
    </p>
  </div>
</div>
`;
